lib/galaxy/jobs/runners/drmaauniva.py

```diff
@@ -0,0 +1,640 @@
+        elif state in [ self.drmaa.JobState.QUEUED_ACTIVE, self.drmaa.JobState.SYSTEM_ON_HOLD, self.drmaa.JobState.USER_ON_HOLD, self.drmaa.JobState.USER_SYSTEM_ON_HOLD, self.drmaa.JobState.RUNNING, self.drmaa.JobState.SYSTEM_SUSPENDED, self.drmaa.JobState.USER_SUSPENDED ]:
+            log.warning( '({tag}/{jobid}) Job is {state}, returning to monitor queue'.format(tag=ajs.job_wrapper.get_id_tag(), jobid=ajs.job_id, state=self.drmaa_job_state_strings[state] ))
+            # TODO return True?
+            return True  # job was not actually terminal
+        else:
+            logging.error( "DRMAAUniva: job {job_id} determined unknown state {state}".format(job_id=ajs.job_id, state=state) )
+            drmaa_state = self.drmaa_job_states.FAILED
+        # by default, finish the job with the state from drmaa
+        return super( DRMAAUnivaJobRunner, self )._complete_terminal_job( ajs, drmaa_state=drmaa_state )
+
+    def _drmaa_state_is_refined( self, statep, staten ):
+        """
+        check if staten is more 'severe' than statep
+        """
+        # definition of the severity of job states, the hex codes are
+        # from the drmaa C library which seem to define a useful order
+        drmaa_job_state_order = {
+            self.drmaa.JobState.UNDETERMINED: 0x00,
+            self.drmaa.JobState.QUEUED_ACTIVE: 0x10,
+            self.drmaa.JobState.SYSTEM_ON_HOLD: 0x11,
+            self.drmaa.JobState.USER_ON_HOLD: 0x12,
+            self.drmaa.JobState.USER_SYSTEM_ON_HOLD: 0x13,
+            self.drmaa.JobState.RUNNING: 0x20,
+            self.drmaa.JobState.SYSTEM_SUSPENDED: 0x21,
+            self.drmaa.JobState.USER_SUSPENDED: 0x22,
+            self.drmaa.JobState.DONE: 0x30,
+            self.drmaa.JobState.FAILED: 0x40 }
+        return drmaa_job_state_order[ staten ] > drmaa_job_state_order[ statep ]
+
+    def _get_drmaa_state_qstat( self, job_id, extinfo=set()):
+        """
+        get a (drmaa) job state with qstat. qstat only returns infos for jobs that
+        are queued, suspended, ..., or just finished (i.e. jobs are still
+        in the system).
+        information on finished jobs can only be found by qacct.
+        Hence if qstat does not contain information on the job
+        the state is assumed as DONE
+        job_id the job id
+        extinfo a set that additional information can be stored in, i.e., "deleted"
+        returns the drmaa state
+        """
+        logging.debug("DRMAAUnivaJobRunner._get_drmaa_state_qstat ({jobid})".format(jobid=job_id))
+        # TODO using -u "*" is the simplest way to query the jobs of all users which
+        # allows to treat the case where jobs are submitted as real user it would
+        # be more efficient to specify the user (or in case that the galaxy user
+        # submits the job -> to ommit the -u option)
+        cmd = ['qstat', '-u', '"*"']
+        p = subprocess.Popen( cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE )
```

Comment:
To avoid duplication these should probably be reduced to a single utility function (I should do this in the Slurm runner as well!). Maybe `check_output_error` mimicking the behavior of `subprocess.check_output` (which there's a copy of [here](https://github.com/galaxyproject/starforge/blob/f1013ddf6119381ffba6df06e8d996cee5678280/starforge/util.py#L43)), just return the stderr with it as a tuple, e.g. `return (output, error)`

Reply:
Hmm. Git the idea and started coding, but was **not entirely convinced** that it is useful since the code in my python class gets larger... 

The reason is that I often check stderr/errorcode quite a bit. So I would need a function that returns stdout, stderr, and the errorcode. Furthermore the function can not raise an exception if stderr!="" or errorcode because I still need these outside of the function. 

Hence I would only replace two lines (Popen and communicate) by one (the new function call).